File: c_examples/byte_rep.c
#include <stdio.h>

// Define a type for a pointer to an unsigned char
// typedef is used to create an alias for a data type. In this case, 'byte_pointer' is an alias for 'unsigned char *'.

typedef unsigned char *byte_pointer;

// Function to print bytes of any data type
void show_bytes(byte_pointer start, size_t len) {
    int i;
    for (i = 0; i < len; i++) {
        printf(" %.2x", start[i]); // Print each byte as two hexadecimal digits
    }
    printf("\n");
}

// Function to show bytes of an integer
void show_int(int x) {
    show_bytes((byte_pointer) &x, sizeof(int)); // Cast integer address to byte_pointer
}

// Function to show bytes of a float
void show_float(float x) {
    show_bytes((byte_pointer) &x, sizeof(float)); // Cast float address to byte_pointer
}

// Function to show bytes of a pointer
void show_pointer(void *x) {
    show_bytes((byte_pointer) &x, sizeof(void *)); // Cast pointer address to byte_pointer
}

// Test function to demonstrate the byte representation of different data types
void test_show_bytes(int val) {
    int ival = val; // Integer variable
    float fval = (float)ival; // Float variable initialized from int
    int *pval = &ival; // Pointer to the integer variable
    show_int(ival); // Show bytes of integer
    show_float(fval); // Show bytes of float
    show_pointer(pval); // Show bytes of pointer
}

// Main function to run the test
int main() {
    test_show_bytes(12345); // Test with the integer value 12345
    return 0;
}

File: rust_examples/src/byte_rep.rs
// Define a type for a pointer to an unsigned char
type BytePointer<'a> = &'a [u8];

// Function to print bytes of any data type
fn show_bytes(start: BytePointer) {
    for byte in start {
        print!(" {:02x}", byte);
    }
    println!();
}

// Function to show bytes of an integer
fn show_int(x: i32) {
    let bytes = x.to_ne_bytes();
    show_bytes(&bytes);
}

// Function to show bytes of a float
fn show_float(x: f32) {
    let bytes = x.to_ne_bytes();
    show_bytes(&bytes);
}

// Function to show bytes of a pointer
fn show_pointer<T>(x: &T) {
    let pointer: usize = x as *const T as usize;
    let bytes = pointer.to_ne_bytes();
    show_bytes(&bytes);
}

// Test function to demonstrate the byte representation of different data types
pub fn test_show_bytes(val: i32) {
    let ival = val; // Integer variable
    let fval = ival as f32; // Float variable initialized from int
    let pval = &ival; // Pointer to the integer variable
    show_int(ival); // Show bytes of integer
    show_float(fval); // Show bytes of float
    show_pointer(&pval); // Show bytes of pointer
}